### PR TITLE
add testing with envtest

### PR DESCRIPTION
**Pull request by tommyknows**
This commit adds testing! 🥳

It uses the envtest package (which means that etcd and kube-apiserver
must be installed) and wraps the usual phases with testing. This way,
correct behaviour of each phase can be ensured.

Signed-off-by: Ramon Rüttimann <ramon@nine.ch>

### CODE REVIEW

File: migration/migration_test.go [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+
+	if *pvc.Spec.StorageClassName != testNewStorageClass {
+		return errors.Errorf("New PVC's storage Class is incorrect: %v", *pvc.Spec.StorageClassName)
+	}
+
+	return nil
+}
+
+// assertMigrateData ensures that a job is created, creates a dummy pod with that job as
+// an owner, sets the pod's status to succeeded and the job's status to succeeded too.
```
Comment (ctrox) [inline migration/migration_test.go:285]: I don't think you actually need to create the dummy pod here, the `job.Status.Succeeded = 1` should be enough, no?

File: migration/migration_test.go [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+	}
+	if job == nil {
+		return errors.Errorf("no job for PVC %v found", ts.pvc)
+	}
+
+	specTmpl := newJob(ts.pvc, withMigrationSuffix(ts.pvc), ts.namespace).Spec.Template.Spec
+
+	pod := &core.Pod{
+		ObjectMeta: meta.ObjectMeta{
+			Name:      "random-name-" + uniqueID(),
```
Comment (ctrox) [inline migration/migration_test.go:307]: You could just use the `GenerateName` field to let the API server assign a unique name.

File: migration/migration_test.go (hunk 1 above)
Comment (tommyknows) [reply migration/migration_test.go:285]: Technically, yes. But creating the pod here is used to assure that the `cleanup` stage works too & removes the pod again.
I'll add a comment to describe this.

**Comment by tommyknows** [comment]
> Quite complex testing all of this. Especially the `addAssertions` bit, but looks really solid in general!

Kinda agree, that's why there's a rather lengthy comment on that function. And I don't think it's a lot more complex than the idea (e.g. the complexity between theory and practice is roughly equal). 

Do you think it needs a better description? Or could be simplified somehow?

File: migration/migration_test.go (hunk 1 above)
Comment (ctrox) [reply migration/migration_test.go:285]: Sounds good :+1:

**Comment by ctrox** [comment]
> > Quite complex testing all of this. Especially the `addAssertions` bit, but looks really solid in general!
> 
> Kinda agree, that's why there's a rather lengthy comment on that function. And I don't think it's a lot more complex than the idea (e.g. the complexity between theory and practice is roughly equal).
> 
> But I'm probably biased here because I wrote it, so:
> Do you think it needs a better description? Or could be simplified somehow?

No I think it is fine, just wanted to mention it :) The comment explains it pretty well.

**Comment by tommyknows** [comment]
Alright, I fixed the points you mentioned 👍

**Pull request merged**